### PR TITLE
Gh #49 Added implementation for generating extended child keys based on BIP32

### LINKED ISSUE #6: Ability to generate hierarchical keypairs

### LINKED ISSUE #7: Derive private child key from private parent key

### LINKED ISSUE #11: Test the child key derivation
https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#test-vectors 
It should work like this.

### LINKED ISSUE #49: Make the functionality for deriving child keys
Figure out if we are going to make derivation from the extended keys or from the master keys.

### PR DESCRIPTION
**Pull request by gspasov**
resolves #49 
resolves #11
resolves #7
resolves #6

The derivation of child public key from parent public key is not working for now. Will be done some time in the future. An issue has beed created for this problem #65

### CODE REVIEW

**Comment by thepiwo** [comment]
@gspasov can you explain the usage of the mersenne prime here?

**Comment by gspasov** [comment]
## Why the mersenne number

Each extended key has 2^31 normal child keys, and 2^31 hardened child keys. Each of these child keys has an index. The normal child keys use indices 0 through 2^31-1. The hardened child keys use indices 2^31 through 2^32-1. To ease notation for hardened key indices, a number iH represents i+2^31.


## Derivation

Given a parent extended key and an index i, it is possible to compute the corresponding child extended key. The algorithm to do so depends on whether the child is a hardened key or not (or, equivalently, whether i ≥ 2^31), and whether we're talking about private or public keys.


## Additionally

Somewhat more care must be taken regarding extended keys, as these correspond to an entire (sub)tree of keys.

One weakness that may not be immediately obvious, is that knowledge of a parent extended public key plus any non-hardened private key descending from it is equivalent to knowing the parent extended private key (and thus every private and public key descending from it). This means that extended public keys must be treated more carefully than regular public keys. It is also the reason for the existence of hardened keys, and why they are used for the account level in the tree. This way, a leak of account-specific (or below) private key never risks compromising the master or other accounts.

From [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki)

**Comment by cytadela8** [comment]
@gspasov Oh, so it's more about the number being `2^31 - 1` not about being a prime number. Ok.

File: lib/elixir_wallet/elixir_wallet.ex [hunk 1]
```diff
@@ -60,24 +60,24 @@ defmodule Wallet do
...
           pass_phrase_check = Enum.at(mnemonic_list, 12)
-        case pass_phrase_check do 
+        case pass_phrase_check do
           :nil -> {:ok , mnemonic}
-          result -> 
-            if result == pass_phrase do 
-              mnemonic = String.replace(mnemonic, " " <> pass_phrase, "")          
-              {:ok, mnemonic} 
-            else 
+          result ->
```
Comment (meivantodorov) [inline lib/elixir_wallet/elixir_wallet.ex]: You can use guard here instead of `if do`

File: test/child_key_derivation.ex [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  #    master_key
+  #    |> KeyPair.derive("m/0'/1/2'/2")
+  #    |> KeyPair.generate_master_key(:public)
+  #  public_child_key = public_master_key |> KeyPair.derive("m/1000000000")
+
+    ## Chain m/0h/1/2h/2/1000000000
+  #  assert "xpub6H1LXWLaKsWFhvm6RVpEL9P4KfRZSW7abD2ttkWP3SSQvnyA8FSVqNTEcYFgJS2UaFcxupHiYkro49S8yGasTvXEYBVPamhGW6cFJodrTHy"
+  #  = KeyPair.format_key(public_child_key)
+  #end
+end
```
Comment (meivantodorov) [inline test/child_key_derivation.ex]: Maybe a good explanation why You keep this commented code here will be nice. `Not working` its not enough :)

**Comment by gspasov** [comment]
Will fix the issues now :)

**Pull request merged**